[8팀 박창준] Chapter 2-1. 클린코드와 리팩토링

ckdwns9121 opened the pull request:
## 과제 체크포인트

### 배포링크

[배포링크](https://devchangjun.github.io/front_6th_chapter2-1/)

### 기본과제

- [x] 코드가 Prettier를 통해 일관된 포맷팅이 적용되어 있는가?
- [x] 적절한 줄바꿈과 주석을 사용하여 코드의 논리적 단위를 명확히 구분했는가?
- [x] 변수명과 함수명이 그 역할을 명확히 나타내며, 일관된 네이밍 규칙을 따르는가?
- [x] 매직 넘버와 문자열을 의미 있는 상수로 추출했는가?
- [x] 중복 코드를 제거하고 재사용 가능한 형태로 리팩토링했는가?
- [x] 함수가 단일 책임 원칙을 따르며, 한 가지 작업만 수행하는가?
- [x] 조건문과 반복문이 간결하고 명확한가? 복잡한 조건을 함수로 추출했는가?
- [x] 코드의 배치가 의존성과 실행 흐름에 따라 논리적으로 구성되어 있는가?
- [x] 연관된 코드를 의미 있는 함수나 모듈로 그룹화했는가?
- [x] ES6+ 문법을 활용하여 코드를 더 간결하고 명확하게 작성했는가?
- [x] 전역 상태와 부수 효과(side effects)를 최소화했는가?
- [x] 에러 처리와 예외 상황을 명확히 고려하고 처리했는가?
- [x] 코드 자체가 자기 문서화되어 있어, 주석 없이도 의도를 파악할 수 있는가?
- [x] 비즈니스 로직과 UI 로직이 적절히 분리되어 있는가?
- [x] 코드의 각 부분이 테스트 가능하도록 구조화되어 있는가?
- [x] 성능 개선을 위해 불필요한 연산이나 렌더링을 제거했는가?
- [x] 새로운 기능 추가나 변경이 기존 코드에 미치는 영향을 최소화했는가?
- [x] 코드 리뷰를 통해 다른 개발자들의 피드백을 반영하고 개선했는가?
- [x] (핵심!) 리팩토링 시 기존 기능을 그대로 유지하면서 점진적으로 개선했는가?

### 심화과제

- [x] 변경한 구조와 코드가 기존의 코드보다 가독성이 높고 이해하기 쉬운가?
- [x] 변경한 구조와 코드가 기존의 코드보다 기능을 수정하거나 확장하기에 용이한가?
- [x] 변경한 구조와 코드가 기존의 코드보다 테스트를 하기에 더 용이한가?
- [x] 변경한 구조와 코드가 기존의 모든 기능은 그대로 유지했는가?
- [x] (핵심!) 변경한 구조와 코드를 새로운 한번에 새로만들지 않고 점진적으로 개선했는가?


## 과제 셀프회고

우선 과제 양이 생각보다 많아서 놀랐습니다. 다시 1주차 과제로 돌아간 느낌이 들기도 했네요...
하지만 제가 평소에 레거시 코드를 리팩토링하고 클린한 구조로 개선하는 것에 관심이 많다 보니, 이번 과제는 꽤 즐겁게 몰입해서 할 수 있었습니다.

다만 코드를 확인해보니 **아 이거 쉽지 않겠구나..** 싶었습니다.  AI 시대에 있어서 AI가 코드 생산은 잘 해주지만, 이런 **복잡하고 엉켜버린 레거시 코드를 깔끔한 아키텍쳐로 재정비 하는 일은 AI가 할 수 있을까? 이건 사람의 영역 아닌가?** 하는 궁금증도 들었고 거의 재건축 수준의 작업이 될 것 같았습니다.

도대체 어디서부터 손을 대야하지? 하는 막막함이 가장 컸는데 AI에게 어떤 순서대로 리팩토링을 하면 좋을지 조언을 받았고, 그걸 기반으로 제 나름의 리팩토링 플로우를 정립해봤습니다.

변수 선언 개선 (var 제거, let to const) → 중복 변수 제거 → 매직넘버 상수화 → 관심사 분리  → 함수 분리 및 단일 책임 적용 → 모듈 화 및 구조 개선 → 고급 리팩토링(디자인패턴 적용이라던지, 새로운 아키텍쳐 적용)

이런 흐름으로 작업을 진행했습니다.

그리고 도은님이 추천해주신 Cursor Rules를 기반으로 클린코드 룰도 적용하여서 많은 도움을 받았습니다. 감사합니다.

### 과제를 하면서 내가 제일 신경 쓴 부분은 무엇인가요?

### **1. 단일 책임 원칙을 지키기 위한 구조 설계**

- 비즈니스 로직, UI 렌더링, 상태 저장, 데이터 변환 등 각각의 역할이 섞이지 않도록 최대한 명확한 경계로 나누는 것을 최우선으로 했습니다.
- 이를 위해 **UI에서 비즈니스 상태를 수정하지 않고**, **비즈니스 로직에서 DOM에 직접 접근하지 않는** 원칙을 지키면서 코드를 수정했습니다.
- 비즈니스 로직은 Service단, 상태관리는 Store단, 이벤트 관리는 Lisnter단에서 관리하였습니다.

```jsx
// Service단에서는 비즈니스 로직을 처리하게 함.

export class CartService {
  addProductToCart(product, quantity = 1) {
    if (!this.validateStock(product, quantity)) {
      return false;
    }

    const { cartItems } = this.cartStore.getState();
    const existingItem = cartItems.find(item => item.id === product.id);

    if (existingItem) {
      existingItem.quantity += quantity;
    } else {
      const newItem = {
        id: product.id,
        name: product.name,
        price: product.price,
        quantity,
      };
      this.cartStore.setState({
        cartItems: [...cartItems, newItem],
      });
    }

    product.quantity -= quantity;
    this.updateCartTotals();
    return true;
  }
}

// Store는 상태와, getter setter만 구현해서 단일 책임을 지게함.
export class CartStore {
  constructor() {
    this.state = {
      cartItems: [],
      totalAmount: 0,
      itemCount: 0,
      discountRate: 0,
      savedAmount: 0,
      lastSelectedProduct: null,
    };
  }

  // 불변성을 유지하는 상태 업데이트
  setState(newState) {
    this.state = { ...this.state, ...newState };
  }

  // 상태 조회
  getState() {
    return this.state;
  }
}

export default CartStore;

```

### **2. 순수 함수와 상태 캡슐화**

- 가능한 많은 함수들을 순수 함수로 만들려 했습니다.
- 외부 상태를 참조하거나 변경하지 않고, 동일한 입력에 대해 동일한 출력을 보장하도록 설계하려고 신경을 썼습니다.
- 또한 상태는 `store`로 캡슐화를 진행하였습니다.

```jsx
// ❌ 부작용이 있는 함수
function calculateDiscount(cartItems) {
  // 외부 상태 참조
  const tuesdayDiscount = isTuesday(); // 외부 함수 호출

  // 외부 상태 변경
  document.querySelector("#discount").textContent = discount;

  return discount;
}

// ✅ 순수 함수
export function calculateCartSummary(cartItems, discountService, cartService) {
  // 모든 의존성을 매개변수로 받음
  const discountResult = discountService.applyAllDiscounts(cartItems, productList);
  const itemCount = cartService.getItemCount();

  // 외부 상태 변경 없음
  return {
    success: true,
    cartItems,
    discountResult,
    itemCount,
  };
}

// ✅ 상태 캡슐화
export class CartService {
  constructor() {
    this.cartStore = new CartStore(); // 상태 관리 캡슐화
  }

  getState() {
    return this.cartStore.getState(); // 상태 접근 제어
  }

  setState(newState) {
    this.cartStore.setState(newState); // 상태 변경 제어
  }
}

```

### **3. 서비스 간 의존성 최소화**

- 순환 의존성을 방지하려고 **의존성 주입 패턴**을 적용하였습니다.
- 예를들어서 `CartService`는 `DiscountService`와 `ProductService`를 외부에서 주입받도록 설계하여 서비스의 의존성을 최소화 하였습니다.

```jsx
// ❌ 강한 의존성
class CartService {
  constructor() {
    this.discountService = new DiscountService(); // 직접 생성
    this.productService = new ProductService();
  }
}

// ✅ 의존성 주입
export class CartEventListeners {
  constructor(uiEventBus, cartService, discountService, productService) {
    this.uiEventBus = uiEventBus;
    this.cartService = cartService;
    this.discountService = discountService;
    this.productService = productService;
  }
}

// 메인에서 의존성 주입
const cartService = new CartService();
const discountService = new DiscountService();
const productService = new ProductService();

const cartListeners = new CartEventListeners(uiEventBus, cartService, discountService, productService);

```

### **4. Event Bus 아키텍처 도입**

이번 과제에서 가장 큰 고민을 했던 부분은 **비즈니스 로직**과 **UI 로직의 분리**였습니다. 이 고민을 해결하려고 이벤트 버스(EventBus) 아키텍처를 도입해 모듈 간 결합도를 낮추는 방식으로 접근했는데요, 구조 자체는 느슨한 연결을 만들어 유지보수에 도움이 되었습니다.

```jsx

    // 장바구니 요약 계산 완료 이벤트 처리
    this.uiEventBus.on(CART_SUMMARY_CALCULATED, data => {
      if (data.success) {
        this.renderCartUI(data.cartItems, data.discountResult, data.itemCount);
      }
    });

    // 장바구니 아이템 스타일 업데이트 이벤트
    this.uiEventBus.on(CART_ITEM_STYLES_UPDATED, data => {
      if (data.success) {
        this.updateCartItemStyles(data.cartItems);
      }
    });

    // 장바구니 가격 업데이트 요청 이벤트 처리
    this.uiEventBus.on(CART_PRICES_UPDATE_REQUESTED, data => {
      if (data.success) {
        this.handleCartPricesUpdate();
      }
    });

    // 헤더 아이템 카운트 업데이트 이벤트
    this.uiEventBus.on(HEADER_ITEM_COUNT_UPDATED, data => {
      if (data.success) {
        this.renderHeaderItemCount(data.itemCount);
      }
    });
```

하지만 구조적인 한계도 분명히 경험하게 되었습니다.  실제로는 아래와 같은 **구조적 한계와 설계 고민**이 함께 따랐습니다.

- 어떤 레이어에 이벤트버스를 위치시킬지?
- 이벤트 버스를 싱글톤으로 둘지, 모듈마다 주입을 시킬지
- 특히나 어려웠던 이벤트 키 네이밍…

그렇지만 분명 구조적인 장점이 있었고 저는 두 가지를 뽑아봤습니다.

**느슨한 결합**

이벤트 버스를 중심으로 모듈 간 직접 호출 없이 메시지를 주고받는 구조로 변경하면서, 모듈 간 **의존성을 낮추고 변경에 유연한 구조**를 만들 수 있었습니다. 예를 들어 `ProductSelector`는 `CartService`를 직접 알지 못하지만, `CART_ADD_REQUESTED` 이벤트만 발행하면 나머지 처리는 이벤트 흐름 안에서 자연스럽게 이어집니다.

**단일 책임**

기존에는 비즈니스 로직, 상태 관리, 렌더링 로직이 한 파일에 뒤섞여 있었지만, 이를 `CartService`, `CartStore`, `CartList`, `CartEventListeners` 등으로 나누어 각 모듈이 **하나의 책임만 갖도록 구조화**했습니다. 유지보수는 물론, 의도를 파악하기에도 훨씬 쉬워졌습니다.

아래는 위 구조를 시각화 해보았습니다.

```mermaid
graph TB
    subgraph "Loose Coupling (느슨한 결합)"
        A[ProductSelector] -->|CART_ADD_REQUESTED| B[EventBus]
        B -->|handleAddToCart| C[CartEventListeners]
        C -->|addProductToCart| D[CartService]
        D -->|setState| E[CartStore]
        C -->|CART_ITEM_ADDED| B
        B -->|createCartItem| F[CartList]
    end
    
    subgraph "Single Responsibility (단일 책임)"
        G[CartService<br/>비즈니스 로직만]
        H[CartEventListeners<br/>이벤트 처리만]
        I[CartStore<br/>상태 관리만]
        J[CartList<br/>UI 렌더링만]
    end
    
    subgraph "Testability (테스트 가능성)"
        K[Unit Test<br/>CartService]
        L[Unit Test<br/>CartStore]
        M[Integration Test<br/>EventBus]
    end

```

### 과제를 다시 해보면 더 잘 할 수 있었겠다 아쉬운 점이 있다면 무엇인가요?

너무 욕심을 많이 내다 보니 아쉬움이 많습니다. 처음부터 컨벤션을 제대로 정하고 폴더구조나 데이터 흐름, UI로직 등을 좀더 고민했더라면 더 좋은 구조가 나왔지 않았을까 생각합니다. 무작정 리팩토링 부터 진행하면서 계획대로 되지 않았고, 그러다보니 구조를 여러번 바꾸고 분리해보면서 **중복되는 로직**이나 **순환 의존성** 같은 문제들이 나타났습니다.

그중에서도 가장 아쉬웠던 선택은 **이벤트 버스패턴 도입**이었습니다. 처음에는 모듈 간 결합도를 낮추고 깔끔한 아키텍처를 만들 수 있을 것이라 기대했지만.. 이벤트 흐름을 추적하기 어렵고, 이벤트 체인이 길어질수록 디버깅 난이도가 급격히 올라가며,특히 네이밍과 이벤트 설계 원칙이 모호해지면서 유지보수성이 떨어지는 문제 등을 경험했습니다. 결국, **옵저버 패턴 + 상태 관리** 정도로만 설계해도 이번 과제의 요구사항에는 충분했을 것 같다는 아쉬움이 남습니다. 

하지만 반대로, 이번 기회에 이렇게 이벤트 버스 패턴을 직접 적용해보면서 덕분에 **이벤트 버스 아키텍처가 어떤 상황에서 효과적이고 어떤 상황에서는 과할 수 있는지**를 실제로 겪고 배울 수 있었습니다.

앞으로는 **복잡한 구조가 ‘잘 만든 구조’는 아니라는 것**, 그리고 **경량화된 아키텍처가 오히려 더 유지보수성과 코드의 가독성을 높여준다는 점** 을 더 고려해서 설계할 수 있을 것 같습니다 ㅎㅎ

두번째로 아쉬운점은.. 비즈니스 로직과, 유틸, 헬퍼 함수를 깔끔하게 분리한뒤 리액트에서 재사용하게 하고싶었습니다. 하지만 생각보다 잘 안되었네요.. 아마 클래스를 사용해서 비즈니스 로직을 관리하다 보니 리액트에서 재사용하기가 어려웠던 것 같습니다 

### 리뷰 받고 싶은 내용이나 궁금한 것에 대한 질문 편하게 남겨주세요 :)

### Q1. 특정 서비스에서 아주 일부 기능(1~2개 함수)만 사용하기 위해 전체 서비스를 주입하는 건 과도한 의존일까요?

예를 들어, 저는 `OrderService`에서 `DiscountService`의 **`calculateProductDiscountRate()` 함수 하나만 사용하기 위해** 전체 `DiscountService`를 주입했습니다:

```jsx
export class OrderService {

  constructor(discountService) {
    this.orderStore = new OrderStore();
    this.discountService = discountService;
  }
   
  calculateOrderData(cartItems,productList){
        // 주입받은 discountService를 사용하여 할인 계산
    const discountResult = this.discountService.applyAllDiscounts(cartItems, productList);
   ... 
  }
  
```

그리고 실제 사용하는건

```jsx
const discountResult = this.discountService.applyAllDiscounts(cartItems, productList);
```

이 한줄입니다.

이처럼 특정 서비스의 일부 기능만 필요할 때, 전체 서비스 객체를 통째로 주입하는 게 맞는지? **단일 책임**이나 **의존성 최소화** 관점에서 보면, 오히려 이 주입이 너무 큰 단위로 이루어졌다고도 느껴졌습니다.

그래서 해당 함수만 따로 유틸로 분리해서 사용하면 어떨까 라는 고민도 했는데 결국 DiscountService(할인 관련된 로직)의 응집도가 깨져버리고 할인 정책에 대한 책임이 이곳저곳 흩어지는 결과를 낳게되는게 아닐까? 라는 고민도 들었습니다..

실무에서는 이런 비슷한 상황이 있으면 어떤 기준으로 설계 결정을 하시는지 궁금합니다.

ckdwns9121 committed:
feat: eslint, prettier 설정정

ckdwns9121 committed:
refactor(cart): 전역 상태를 Shopping Cart 모듈로 임시 리팩토링

- 전역 변수들을 모듈 패턴으로 캡슐화
- 상태 관리를 중앙화하여 예측 가능한 구조로 변경
- DOM 요소들도 모듈을 통해 관리
- 모든 함수들이 ShoppingCart 모듈을 사용하도록 수정
- Linter 오류 수정 (중복 변수, 빈 블록)

ckdwns9121 committed:
chore: reset

ckdwns9121 committed:
fix: eslint 설정 수정

ckdwns9121 committed:
feat: eslint 설정 추가

ckdwns9121 committed:
refactor: var 변수 제거 및 불필요한 로직 삭제

ckdwns9121 committed:
refactor: 매직넘버 상수화

ckdwns9121 committed:
refactor: 매직넘버 리팩토링

ckdwns9121 committed:
refactor: 헤더 컴포넌트 분리

ckdwns9121 committed:
refactor: ProductSelector 컴포넌트 분리

- stockInfo undefined 에러 해결.

ckdwns9121 committed:
refactor: handleAddToCart 함수 복잡도 최적화.

- 함수를 4개의 작은 함수로 분리하여 복잡도 최적화
- validateSelectedItem, incrementExistingItem, createNewCartItem 추가
- 단일 책임 원칙 적용으로 코드 품질 향상

ckdwns9121 committed:
refactor: handleAddToCart 함수 변수명 Clean Code 규칙 적용

- 변수명을 더 명확하고 일관성 있게 개선
- 축약어 제거 (qty → quantity, disp → display)
- 의미있는 이름으로 변경 (hasItem → isProductExists)
- DOM 요소임을 명시 (cartItem → cartItemElement)

ckdwns9121 committed:
feat: CartItem 컴포넌트 분리

- 장바구니 아이템 카드 컴포넌트 생성
- 수량 조절, 제거, 가격 업데이트 기능 포함
- 불필요한 createNewCartItem 함수 제거
- 이벤트 처리를 컴포넌트 내부로 이동

ckdwns9121 committed:
feat: 상품 목록 및 장바구니 기능 개선

- prodList를 상수로 정의하여 상품 정보를 명확하게 관리
- validateSelectedItem, incrementExistingItem, createAndAddCartItem, handleQuantityChange, handleRemoveItem 함수 추가로 장바구니 기능 개선
- handleAddToCart 함수에서 selectorContainer 인자 제거 및 코드 간소화
- ProductSelector 컴포넌트의 props 구조 개선
- 코드 스타일 통일성 유지 및 가독성 향상

ckdwns9121 committed:
refactor: 상품 정보 및 장바구니 로직 개선

- 상품 ID 및 가격 관련 변수명을 일관되게 변경 (val → price, q → quantity)
- 장바구니 관련 함수에서 변수명 및 로직을 개선하여 가독성 향상
- CartItem 및 ProductSelector 컴포넌트에서 가격 표시 방식 통일
- 코드 스타일 통일성 유지 및 불필요한 줄 제거

ckdwns9121 committed:
feat: OrderSummary 컴포넌트 추가 및 기존 컴포넌트 리팩토링

- OrderSummary 컴포넌트를 새로 생성하여 주문 요약 정보를 관리
- CartItem, ProductSelector, Header 컴포넌트에서 props 구조 개선 및 불필요한 코드 제거
- createHeader, createProductSelector, createCartItem 함수에서 props 구조를 간소화
- 전체 코드 스타일 통일성 유지 및 가독성 향상
- index.js 파일 삭제 및 각 컴포넌트에서 직접 import/export 처리

ckdwns9121 committed:
fix: 장바구니 비어있을 때 숨김김

ckdwns9121 committed:
refactor: 상품 관련 상수 및 컴포넌트 구조 개선

- 상품 정보를 PRODUCT_LIST로 통합하여 관리
- 기존의 prodList를 제거하고, 관련 함수에서 PRODUCT_LIST 사용으로 변경
- CartItem, Header, OrderSummary, ProductSelector 컴포넌트에서 props 구조 개선 및 불필요한 코드 제거
- 새로운 상수 파일 추가로 코드 가독성 및 유지보수성 향상
- 전체 코드 스타일 통일성 유지

ckdwns9121 committed:
feat: cart 상태를 store와 action으로 분리

ckdwns9121 committed:
feat: cart Store 및 Action으로 분리

ckdwns9121 committed:
chore: 불필요한 변수 제거

ckdwns9121 committed:
feat: cartService를 도입하여 장바구니 로직 개선

- cartService 클래스를 추가하여 장바구니 관련 비즈니스 로직을 분리
- handleAddToCart, handleQuantityChange, handleRemoveItem 함수에서 cartService의 메서드 사용
- 장바구니 아이템 추가, 수량 변경 및 제거 시 재고 검증 로직 통합
- 코드 가독성 및 유지보수성 향상

ckdwns9121 committed:
refactor: 테스트 코드 요일 수정

ckdwns9121 committed:
feat: 장바구니 계산 및 할인 로직 개선

- 장바구니 총계 계산 및 할인 적용 로직을 새로운 유틸리티 함수로 분리
- 재고 표시 경고 기준 추가 및 관련 로직 수정
- 상품 ID로 제품을 찾는 기능을 유틸리티로 분리하여 코드 재사용성 향상
- UI 업데이트 함수 추가로 코드 가독성 및 유지보수성 향상

ckdwns9121 committed:
feat: 레이아웃 및 매뉴얼 시스템 컴포넌트 추가

- 그리드 레이아웃을 생성하는 createGridContainer, createLeftColumn, createRightColumn 함수 추가
- 매뉴얼 토글 버튼, 오버레이, 컬럼을 생성하는 createManualToggle, createManualOverlay, createManualColumn 함수 추가
- 매뉴얼 시스템을 구성하고 이벤트 핸들러를 설정하는 createManualSystem 함수 추가
- OrderSummary 컴포넌트에서 상수들을 외부로 분리하여 코드 가독성 향상

ckdwns9121 committed:
feat: 애플리케이션 전역 상태 관리 클래스 추가

- AppState 클래스를 새로 생성하여 애플리케이션의 전역 상태를 관리
- UI 참조, 장바구니 상태, 상태 변경 리스너 관리 기능 포함
- 상태 업데이트 및 조회 메서드 추가로 코드 구조 개선

ckdwns9121 committed:
feat: 장바구니 및 UI 관련 유틸리티 함수 추가 및 코드 정리

- DOM 조작을 위한 유틸리티 함수(getCartItemQuantity, setCartItemQuantity, extractNumberFromText) 추가
- 재고 관련 유틸리티 함수(calculateTotalStock, generateStockWarningMessage, getLowStockProducts) 추가
- main.basic.js 파일에서 전역 상태 관리 및 UI 업데이트 로직 개선
- 불필요한 코드 제거 및 함수 호출 통일성 향상

ckdwns9121 committed:
refactor: 장바구니 계산 및 UI 업데이트 로직 개선

- handleCalculateCartStuff 함수 제거 및 updateCartSummary 함수로 통합
- cartTotals 및 discountResult를 활용하여 상태 업데이트 및 UI 갱신 로직 개선
- subtotal 변수로 이름 변경하여 코드 가독성 향상
- main.basic.js 및 cartCalculations.js 파일에서 관련 코드 정리

ckdwns9121 committed:
feat: 타이머 서비스 추가 및 할인 로직 통합

- 새로운 TimerService 클래스를 추가하여 번개 세일 및 추천 세일 타이머 로직을 관리
- main.basic.js에서 기존의 할인 로직을 TimerService로 이동하여 코드 구조 개선
- 타이머 서비스 초기화 및 시작 로직을 main 함수에 통합하여 가독성 향상
- 관련 상수 및 유틸리티 함수 사용으로 코드 일관성 강화

ckdwns9121 committed:
feat: ProductService 및 ProductStore 추가로 상품 관리 로직 개선

- ProductService 클래스를 추가하여 상품 검증, 재고 관리 및 할인 적용 로직을 통합
- ProductStore 클래스를 도입하여 상품 목록 및 재고 상태를 관리
- main.basic.js에서 상품 관련 로직을 ProductService로 이전하여 코드 구조 개선
- TimerService에서 상품 관련 메서드를 ProductService로 변경하여 의존성 관리 강화

ckdwns9121 committed:
refactor: AppState 클래스 삭제 및 전역 상태 관리 개선

- AppState 클래스를 삭제하여 전역 상태 관리 방식을 개선
- 관련 UI 참조 및 상태 변경 리스너 관리 기능을 다른 구조로 통합
- 코드 간소화 및 유지보수성 향상

ckdwns9121 committed:
refactor: 장바구니 및 UI 업데이트 로직 개선

- cartStore 클래스를 도입하여 장바구니 상태 관리 및 비즈니스 로직을 통합
- eventBus를 추가하여 컴포넌트 간 통신을 위한 이벤트 시스템 구현
- main.basic.js에서 장바구니 관련 함수들을 cartStore 및 eventBus를 활용하여 리팩토링
- TimerService에서 cartService를 사용하여 장바구니 관련 로직을 간소화
- 코드 가독성 및 유지보수성 향상

ckdwns9121 committed:
refactor: cartService 및 이벤트 시스템 정리

- eventBus 관련 코드 제거 및 cartService에서 이벤트 발행 로직 간소화
- setLastSelectedProduct 메서드에서 이벤트 발행 제거
- main.basic.js에서 불필요한 코드 제거 및 가독성 향상
- 전체적인 코드 구조 개선 및 유지보수성 향상

ckdwns9121 committed:
refactor: CartService 및 관련 코드 개선

- CartService에서 cartStore를 CartStore로 변경하여 인스턴스화
- main.basic.js에서 CartService 초기화 및 전역 상태 관리 개선
- 코드 가독성 향상 및 일관성 있는 클래스 사용으로 유지보수성 강화

ckdwns9121 committed:
refactor: 헤더 전역변수 삭제제

ckdwns9121 committed:
refactor: stockInfo 변수 제거 및 관련 함수 수정

- stockInfo 전역 변수를 제거하고, 필요한 함수에서 직접 DOM 요소를 조회하도록 변경
- updateCartSummary 및 updateCartUI 함수에서 stockInfo 인자를 제거하여 코드 간소화
- 전반적인 코드 가독성 및 유지보수성 향상

ckdwns9121 committed:
refactor: stockInfo 요소 제거 및 updateCartSummary 함수 수정

- stockInfo 요소를 제거하고 관련 함수에서 직접 DOM 요소를 조회하도록 변경
- updateCartSummary 함수에서 stockInfo 인자를 제거하여 코드 간소화
- 전반적인 코드 가독성 및 유지보수성 향상

ckdwns9121 committed:
refactor:: cartDisplay 인자 삭제 작업

ckdwns9121 committed:
refactor: updateCartSummary 및 updateProductOptions 함수 인자 수정

- updateCartSummary 함수에서 cartDisplay 인자를 제거하고 selectorContainer로 변경
- updateProductOptions 함수에서 container 인자를 제거하고 products로 변경
- 전반적인 코드 간소화 및 가독성 향상

ckdwns9121 committed:
refactor: container 인자 삭제

ckdwns9121 committed:
refactor: selectorContainer 인자 제거 및 관련 함수 수정

ckdwns9121 committed:
refactor: calculateDiscountRate 유틸 함수 통합

- CartItem과 ProductSelector에 동적 할인율 계산 추가
- 하드코딩된 할인율을 계산된 값으로 대체
- 정확한 할인 정보로 사용자 경험 개선

ckdwns9121 committed:
refactor: 할인 상태 계산 및 관련 함수 수정

- CartItem 및 ProductSelector에서 할인 상태를 동적으로 계산하도록 수정
- createCartItem 및 getOptionText 함수에서 할인 상태를 반영하여 사용자 경험 개선
- 불필요한 코드 제거 및 가독성 향상

ckdwns9121 committed:
refactor: update cart item handling and simplify price update logic

ckdwns9121 committed:
refactor: remove onProductSelect parameter from createProductSelector function

ckdwns9121 committed:
refactor: integrate orderService for order summary and points calculation

ckdwns9121 committed:
refactor: 컴포넌트에서 discountService 직접 의존성 제거

- CartItem, ProductSelector에서 discountService 직접 호출 제거
- 할인 정보를 props로 전달하여 컴포넌트 순수성 확보
- main.basic.js에서 할인 계산 후 컴포넌트에 전달

ckdwns9121 committed:
refactor: enhance order summary component with subscription to orderService updates

ckdwns9121 committed:
refactor: 포인트 도메인 분리 및 PointService 도입

- PointService 생성하여 포인트 계산 로직 분리
- OrderService에서 calculateLoyaltyPoints 메서드 제거
- CartStore에서 bonusPoints 관련 로직 제거
- PointStore는 불필요하여 제거 (OrderStore에서 포인트 상태 관리)
- 단일 책임 원칙 적용으로 도메인 분리 개선

ckdwns9121 committed:
refactor: cartService와 cartStore에서 사용되지 않는 메서드 제거

- cartService에서 5개 불필요한 메서드 제거
- cartStore에서 구독자 패턴 및 3개 메서드 제거

ckdwns9121 committed:
feat: Event Bus 패턴 구현으로 UI 로직 분리

- 중앙 이벤트 관리 시스템 추가
- 도메인별 이벤트 리스너 분리 (Cart, Product)
- 이벤트 기반 UI 업데이트로 리팩토링
- 포인트 계산 버그 수정

eveneul commented:
배껴요~

ckdwns9121 committed:
refactor: UI 이벤트 버스 및 이벤트 리스너 추가로 관심사 분리

- Cart 및 Product 관련 이벤트 리스너 구현
- DOM 조작을 이벤트 기반으로 변경하여 코드 간결화
- 불필요한 DOM 관련 파일 삭제
- 전역 상태 관리 및 UI 업데이트 로직 개선

ckdwns9121 committed:
refactor: 상품 및 장바구니 업데이트 로직 개선

- 상품 옵션, 재고, 가격 업데이트를 위한 핸들러 함수 추가
- UI 이벤트 버스를 통해 상품 데이터와 상태를 관리하도록 리팩토링
- DOM 조작을 이벤트 기반으로 변경하여 코드 간결화 및 유지보수성 향상

ckdwns9121 committed:
refactor: 장바구니 및 상품 이벤트 처리 로직 개선

- 장바구니 수량 변경 및 아이템 제거 요청을 이벤트 기반으로 처리하도록 리팩토링
- 전역 함수 호출을 제거하고 UI 이벤트 버스를 통해 상태 업데이트
- 상품 추가 로직을 CartEventListeners로 통합하여 코드 간결화 및 유지보수성 향상

ckdwns9121 committed:
refactor: simplify cart and product event handling

ckdwns9121 committed:
refactor: enhance cart and product event handling with event bus integration

ckdwns9121 committed:
refactor: add OrderEventListeners for handling order summary updates

ckdwns9121 committed:
refactor: orderListeners 이벤트 패턴 통일 및 포인트 계산 이벤트 연결 수정

- orderListeners.js를 구독자 패턴에서 event-bus 패턴으로 변경
- 포인트 계산 완료 이벤트 "order:calculation:completed" 추가
- main.basic.js에서 subscribeToChanges 제거하고 event-bus 이벤트 리스너 추가
- 포인트 계산 결과가 UI에 정상 반영되도록 이벤트 연결 수정

ckdwns9121 committed:
refactor: OrderSummary DOM 의존성 제거 및 ID 기반 선택자 적용용

ckdwns9121 committed:
refactor: CartService의 ProductService 의존성 제거

ckdwns9121 committed:
refactor: 이벤트 버스 패턴 개선 - 핸들러 함수 순수화

ckdwns9121 committed:
refactor:  핸들러 순수화 및 중복 로직 제거

- 핸들러 함수들을 순수 비즈니스 로직만 처리하도록 변경
- DOM 직접 접근 제거로 결합도 감소
- 중복된 재고 업데이트 로직을 이벤트 기반 통신으로 통일
- 직접 함수 호출을 이벤트 발송으로 변경

ckdwns9121 committed:
refactor: 불필요한 매직넘버 삭제제

ckdwns9121 committed:
refactor: 불필요한 유틸함수 삭제제

ckdwns9121 committed:
refactor: 핸들러 및 컴포넌트 구조 개선

- 핸들러 함수들을 순수 비즈니스 로직으로 변경하고 DOM 접근을 제거하여 결합도 감소
- 상품 및 재고 관련 핸들러를 통합하여 코드 중복 제거
- OrderSummary 및 ProductSelector 컴포넌트의 DOM 의존성 제거 및 구조 개선
- 이벤트 기반 통신을 통해 장바구니 요약 계산 및 상품 옵션 업데이트 로직 통일

ckdwns9121 committed:
refactor: 핸들러 및 서비스 구조 개선

- 불필요한 핸들러 파일 삭제 및 관련 로직을 서비스로 통합하여 코드 간결화
- 상품 및 장바구니 관련 초기화 로직을 비동기 함수로 변경하여 UI 초기화 개선
- OrderSummary 컴포넌트의 적립 포인트 표시를 숨김 처리하여 UI 개선
- TimerService에서 가격 업데이트 로직을 직접 이벤트 발송으로 변경하여 결합도 감소

ckdwns9121 committed:
refactor: 장바구니 추가 요청 로직 개선

- CartService에서 장바구니 추가 요청을 처리하는 메서드를 간소화하여 이벤트 발송 로직을 직접 호출하도록 변경
- main.basic.js에서 장바구니 추가 요청을 Event Bus 대신 CartService를 통해 처리하도록 수정
- 데이터 파일 PRODUCT_LIST를 추가하여 상품 목록 관리 개선

ckdwns9121 committed:
refactor: 서비스 관리 구조 개선

- ServiceManager 클래스를 추가하여 서비스 인스턴스를 중앙에서 관리하도록 변경
- main.basic.js에서 서비스 초기화 및 이벤트 리스너 등록 방식을 개선하여 코드 간결화
- 각 서비스의 초기화 로직을 ServiceManager를 통해 통합하여 유지보수성 향상

ckdwns9121 committed:
refactor: 서비스 초기화 로직 통합 개선

- ServiceManager의 initializeAll 메서드를 수정하여 모든 서비스의 UI 초기화를 순회하며 처리하도록 변경
- 코드 간결화 및 유지보수성 향상

ckdwns9121 committed:
refactor: 장바구니 할인 계산 로직 개선 및 불필요한 코드 삭제

- CartStore에서 할인 계산 로직을 discountService를 사용하도록 변경하여 코드 간결화
- calculateProductDiscountInfo 메서드를 삭제하고 관련 로직을 통합
- test.js 파일 삭제

ckdwns9121 committed:
refactor: 상품 서비스 및 스토어 코드 정리

- ProductService에서 불필요한 메서드 삭제 및 discountService를 직접 사용하도록 변경하여 코드 간결화
- ProductStore에서 reset 메서드 삭제로 코드 정리

ckdwns9121 committed:
refactor: 주문 요약 및 포인트 계산 로직 통합

- OrderEventListeners에서 주문 요약 및 포인트 계산 로직을 calculateOrderAndPoints 메서드로 통합하여 코드 중복 제거
- OrderService에서 불필요한 메서드 삭제로 코드 간결화

ckdwns9121 committed:
refactor: 상품 서비스 및 타이머 서비스 코드 정리

- ProductService에서 비즈니스 로직 관련 주석 추가 및 불필요한 메서드 삭제로 코드 간결화
- TimerService에서 타이머 로직 개선 및 UI 업데이트 방식을 간소화하여 유지보수성 향상
- 이벤트 발송 로직을 통합하여 코드 중복 제거

ckdwns9121 committed:
chore: 커밋 돌리기 실수

ckdwns9121 committed:
feat: react 설정 추가

ckdwns9121 committed:
refactor: 서비스 구조 개선 및 초기화 로직 삭제제

ckdwns9121 committed:
Merge branch 'main' of https://github.com/devchangjun/front_6th_chapter2-1

ckdwns9121 committed:
refactor: 서비스 관리 구조 변경 및 불필요한 코드 삭제

- ServiceManager 클래스를 core 디렉토리로 이동하여 구조 개선
- DiscountStore에서 구독자 패턴 관련 메서드 및 notifySubscribers 호출 삭제로 코드 간결화

ckdwns9121 committed:
refactor: DiscountStore 및 OrderStore 초기화 방식 변경

- DiscountStore와 OrderStore를 인스턴스화하여 서비스에서 직접 사용하도록 수정
- 불필요한 코드 및 메서드 삭제로 코드 간결화

ckdwns9121 committed:
feat: 장바구니 서비스 및 스토어 기능 개선

- CartService에 재고 검증 및 장바구니 총액 계산 기능 추가
- CartStore에서 불필요한 메서드 삭제 및 상태 관리 개선
- 장바구니 초기화 및 마지막 선택된 상품 설정 기능 추가
- 전체적인 코드 구조 개선으로 유지보수성 향상

ckdwns9121 committed:
refactor: 할인 서비스 및 스토어 상태 관리 개선

- DiscountService에서 할인 집계 로직을 calcTotalDiscount 유틸 함수로 분리하여 코드 가독성 향상
- DiscountStore의 상태 업데이트 방식을 setState 메서드로 통일하여 불변성 유지
- 개별, 대량구매, 화요일 할인 업데이트 로직을 간소화하여 코드 간결화
- 할인 초기화 로직을 개선하여 상태 관리 효율성 증대

ckdwns9121 committed:
refactor: 장바구니 및 주문 서비스 상태 관리 개선

- CartService와 OrderService에서 상태 관리 방식을 setState 메서드로 통일하여 불변성 유지
- CartStore와 OrderStore에서 상태 조회 및 업데이트 메서드 추가로 코드 간결화
- ProductService에서 상품 상태 관리 개선 및 관련 메서드 통합
- 전체적인 코드 구조 개선으로 유지보수성 향상

ckdwns9121 committed:
refactor: 장바구니 추가 요청 방식 변경 및 코드 정리

- CartService의 addToCart 메서드를 삭제하고, UI 이벤트 버스를 통해 장바구니 추가 요청을 처리하도록 변경
- ProductService의 상품 상태 업데이트 로직을 간소화하여 가독성 향상
- DiscountStore에서 불필요한 구독자 관리 메서드 삭제로 코드 간결화

ckdwns9121 committed:
refactor: 장바구니 총액 계산 로직 단순화 및 할인 기능 제거

- updateCartTotals 메서드에서 할인 계산 로직을 제거하고 기본 총액 계산으로 변경
- 할인 관련 상태 및 반환 값 초기화로 코드 간결화
- 전체적인 코드 가독성 향상

ckdwns9121 committed:
refactor: 서비스 및 컴포넌트 구조 개선

- DiscountService를 서비스 매니저에 주입하여 의존성 관리 개선
- CartEventListeners 및 OrderEventListeners에서 UI 업데이트 메서드 이름 변경으로 가독성 향상
- ProductSelector에서 선택된 값 유지 기능 추가로 사용자 경험 개선
- TimerService의 생성자에서 cartDisplay 제거 및 타이머 간격 수정으로 코드 간결화
- OrderSummary에서 cartItems를 포함하여 주문 요약 업데이트 로직 개선

ckdwns9121 committed:
refactor: 상품 할인 정보 계산 로직 추가 및 UI 개선

- ProductSelector에서 할인 정보를 계산하여 전달하도록 수정
- OrderSummary에서 대량구매 할인 문구 변경으로 가독성 향상
- 주문 요약 업데이트 시 orderState 로그 추가로 디버깅 용이성 증대
- OrderService에서 할인 정보 구조 변경으로 코드 일관성 유지

ckdwns9121 committed:
feat: 장바구니 및 주문 관련 컴포넌트 추가

- CartDisplay 컴포넌트 추가로 장바구니 아이템 목록 표시 기능 구현
- CartItem 컴포넌트 추가로 개별 장바구니 아이템의 상세 정보 및 수량 조정 기능 추가
- Header 컴포넌트 추가로 장바구니 아이템 수 표시
- Layout 컴포넌트에 CartDisplay 및 Header 통합
- OrderSummary 컴포넌트 추가로 주문 요약 및 결제 정보 표시 기능 구현
- ProductSelector 컴포넌트 추가로 상품 선택 및 장바구니 추가 기능 구현
- Manual 컴포넌트 추가로 사용자 안내 정보 제공

ckdwns9121 committed:
feat: 상품 관련 상수 및 초기 데이터 추가

- PRODUCT_IDS 상수 추가로 상품 ID 관리 용이성 향상
- INITIAL_PRODUCTS 배열 추가로 초기 상품 데이터 정의
- DISCOUNT_RATES, POINT_RATES, SPECIAL_EVENTS, STOCK_WARNINGS 상수 추가로 할인, 포인트, 이벤트 및 재고 경고 관련 설정 통합

ckdwns9121 committed:
feat: 장바구니 상태 관리 및 기능 구현

- CartContext.tsx 파일 추가로 장바구니 상태 관리 및 관련 기능 구현
- 장바구니 아이템 추가, 수량 업데이트, 제거, 초기화 기능 포함
- 선택된 상품 관리 및 총액, 아이템 수, 보너스 포인트 계산 로직 추가
- React Context API를 사용하여 장바구니 상태 공유 및 관리

ckdwns9121 committed:
feat: 특별 이벤트 훅 추가

- useSpecialEvents 훅을 추가하여 번개세일 및 추천 할인 기능 구현
- 상품 업데이트를 위한 콜백 함수와 타이머 설정 로직 포함
- 사용자에게 할인 알림을 위한 alert 메시지 추가

ckdwns9121 committed:
feat: 할인 및 포인트 계산 서비스 추가

- DiscountService.ts 파일 추가로 개별, 대량 구매, 화요일 할인 계산 기능 구현
- PointService.ts 파일 추가로 기본 포인트, 화요일 보너스, 세트 보너스 및 수량 보너스 계산 기능 구현
- 각 서비스에서 할인 및 포인트 계산 로직을 통합하여 장바구니 기능 강화

ckdwns9121 committed:
feat: TypeScript로 장바구니 및 상품 관리 기능 구현

- main.advanced.tsx 파일 추가로 React 기반의 장바구니 및 상품 관리 애플리케이션 구조 설정
- 상품 초기 데이터 로딩 및 업데이트 기능 포함
- CartContext 및 관련 타입 정의를 위한 index.ts 파일 추가
- 상품, 장바구니, 할인 및 포인트 관련 타입 인터페이스 정의

ckdwns9121 committed:
chore: prettier 재설정정

ckdwns9121 committed:
feat: 상품 할인 정보 통합 및 이벤트 리스너 개선

- CartEventListeners 및 ProductEventListeners에서 할인 정보를 통합하여 상품 데이터 업데이트 시 할인 상태 반영
- TimerService에서 할인 적용된 상품 목록을 사용하여 UI 업데이트 로직 개선
- calculateProductDiscountInfos 메서드에서 할인 정보 계산 로직을 통합하여 코드 일관성 유지
- CartItem 및 ProductSelector 컴포넌트에서 할인 정보 표시 방식 개선

ckdwns9121 committed:
refactor: 상품 선택기에서 가격 및 재고 관련 변수명 변경

- ProductSelector 컴포넌트에서 상품 가격 관련 변수명을 val에서 price로, originalVal에서 originalPrice로 변경
- 재고 관련 변수명을 q에서 quantity로 변경하여 코드 가독성 향상
- 할인 정보 표시 및 재고 부족 메시지 로직 수정

ckdwns9121 committed:
feat: 장바구니 아이템 목록 컴포넌트 추가 및 가격 표시 개선

- CartList 컴포넌트를 새로 추가하여 장바구니 아이템 목록을 표시하도록 구현
- CartItem 컴포넌트에서 가격 관련 변수명을 val에서 price로, originalVal에서 originalPrice로 변경
- 할인 정보 표시 로직을 개선하여 할인율 및 상태를 사용자에게 명확히 전달
- 수량 변경 및 제거 버튼에 데이터 속성 추가로 이벤트 처리 개선

ckdwns9121 committed:
refactor: Layout 및 컴포넌트 구조 개선

- Layout 컴포넌트를 기본 함수형으로 변경하고 CartDisplay를 CartList로 교체하여 장바구니 아이템 목록 표시 방식 개선
- Manual 컴포넌트를 기본 함수형으로 변경하고 버튼에 'manual-close-btn' 클래스 추가
- OrderSummary 컴포넌트에서 가격 관련 변수명을 val에서 price로 변경하여 일관성 유지

ckdwns9121 committed:
refactor: 상품 관련 변수명 및 타입 개선

- 상품 초기 데이터에서 가격 관련 변수명을 val에서 price로, originalVal에서 originalPrice로 변경하여 일관성 유지
- 재고 관련 변수명을 q에서 quantity로 변경하여 가독성 향상
- Product 타입 임포트를 type으로 변경하여 TypeScript의 최적화 적용

ckdwns9121 committed:
refactor: 장바구니 상태 관리에서 변수명 일관성 개선

- 재고 관련 변수명을 q에서 quantity로 변경하여 가독성 향상
- CartProvider 컴포넌트를 함수형으로 변경
- subtotal 계산 시 가격 관련 변수명을 val에서 price로 수정하여 일관성 유지

ckdwns9121 committed:
refactor: 변수명 일관성 및 컴포넌트 구조 개선

- 상품 관련 변수명을 val에서 price로, originalVal에서 originalPrice로 변경하여 일관성 유지
- 재고 관련 변수명을 q에서 quantity로 변경하여 가독성 향상

ckdwns9121 committed:
fix: discountService 싱글톤 인스턴스 제거거

ckdwns9121 committed:
fix: discountService 의존성 주입

ckdwns9121 committed:
refactor: 상품 재고 확인 로직에서 변수명 변경

- hasStock 메서드에서 product 변수를 item으로 변경하여 가독성 향상
- eslint 규칙에 따라 들여쓰기 규칙 비활성화

ckdwns9121 committed:
feat: 로그 추가로 장바구니 및 주문 요약 정보 디버깅 개선

- createCartItemsHTML 함수에 cartItems 로그 추가
- renderOrderSummary 메서드에 orderSummary 및 pointsResult 로그 추가
- OrderService에서 cartItems를 반환하여 주문 정보 처리 개선

ckdwns9121 committed:
refactor: 장바구니 아이템 수량 계산 로직 개선

- OrderService에서 cartItems의 수량 계산 방식을 getCartItemQuantity 함수 사용으로 변경하여 가독성 향상

ckdwns9121 committed:
deploy: advanced 기준 배포 설정

ckdwns9121 committed:
refactor: CartItem 컴포넌트에서 실시간 상품 정보 반영 및 CartList 스타일 개선

- CartItem 컴포넌트에서 useCart 훅을 통해 실시간 상품 정보를 가져와 표시하도록 수정
- 상품 이름, 가격, 할인 정보를 currentProduct를 사용하여 업데이트
- CartList 컴포넌트의 스타일에 space-y-0 클래스를 추가하여 아이템 간의 간격 조정

ckdwns9121 committed:
refactor: Layout 및 OrderSummary 컴포넌트 개선

- Layout 컴포넌트에서 구조를 단순화하고 flexbox를 사용하여 CartList의 스타일을 개선
- OrderSummary 컴포넌트에서 실시간 상품 가격을 반영하도록 subtotal 계산 로직 수정
- CartItems 목록에서 상품 이름과 가격을 실시간으로 업데이트하여 사용자 경험 향상

ckdwns9121 committed:
feat: alert를 toast로 변경경

ckdwns9121 committed:
refactor: 타이밍 값 변경경

ckdwns9121 committed:
refactor: 할인 계산 로직 개선

- 대량 구매 할인 적용 시 개별 상품 할인 제외 로직 추가
- 대량 구매 할인이 없을 때만 개별 상품 할인 적용하도록 수정
- 코드 가독성을 높이기 위해 할인 계산 구조 간소화

ckdwns9121 committed:
feat: toast 알림 추가 및 alert 대체

- 번개세일 및 추천 세일 시 alert 대신 toast 알림 사용
- toast 메시지에 할인 정보 및 지속 시간 설정 추가
- useToast 훅을 통해 toast 기능 통합

ckdwns9121 committed:
refactor: 실시간 상품 가격 반영을 위한 subtotal 계산 로직 수정

- 장바구니의 subtotal 계산 시 현재 상품 정보를 기반으로 최신 가격을 반영하도록 로직 개선
- 상품 가격이 변경될 경우 실시간으로 업데이트되도록 수정하여 사용자 경험 향상

ckdwns9121 committed:
feat: toast 기능 통합 및 App 컴포넌트 구조 개선

- ToastProvider를 추가하여 toast 알림 기능을 App 컴포넌트에 통합
- Layout 컴포넌트와 ToastContainer를 함께 렌더링하여 사용자 경험 향상
- useToast 훅을 통해 toast 메시지 관리 로직 추가

ckdwns9121 committed:
chore: package.json 명령어 수정

ckdwns9121 committed:
fix: update timer intervals

ckdwns9121 committed:
refactor: update createMockCartItem function signature

ckdwns9121 committed:
chore: pnpm lock파일 삭제

ckdwns9121 committed:
refactor: replace CartDisplay with CartList component

ckdwns9121 committed:
chore: 콘솔 삭제

ckdwns9121 committed:
feat: add product and cart related event constants

- Introduced new event constants for product, cart, order, header, stock, and item count updates.
- Updated event listeners in cart and product components to use the new constants for better maintainability.

ckdwns9121 committed:
refactor: update event listeners to use constants for cart and order events

ckdwns9121 committed:
refactor: optimize cart item updates and discount calculations

ckdwns9121 committed:
feat: replace advanced.test.js with advanced.test.tsx for enhanced testing capabilities

ckdwns9121 committed:
feat: TimerService  비즈니스 로직과 UI 로직 분리

ckdwns9121 committed:
fix: TimerService  비즈니스 로직과 UI 로직 분리

ckdwns9121 committed:
refactor: integrate discount calculation into utility function and update event listeners

- Moved the discount calculation logic to a new utility function `calculateProductDiscountInfos`.
- Updated event listeners in `CartEventListeners` and `TimerService` to utilize the new utility function for better code organization.
- Simplified the `createProductSelector` component to handle cart addition events internally.

k-sang-soo commented:
고생하셨습니다 창준님ㅠㅠ

ckdwns9121 committed:
feat: integrate PointService and enhance order processing

ckdwns9121 committed:
refactor:  불필요한 함수 제거

unseoJang commented on src/advanced/components/ProductSelector.tsx:
```diff
@@ -0,0 +1,86 @@
+    }
+  };
+
+  const getProductDisplayText = (product: Product): string => {
```
여기서 let 을 안쓰고 바로 return 시켜도 되지 않을까라는 생각이 드네요
priceText를 하나의 함수로 따로 묶어서
```js
const getPriceText = (): string => {
  if (product.onSale && product.suggestSale) {
    return `⚡💝${product.name} - ${product.originalPrice}원 → ${product.price}원 (25% SUPER SALE!)`;
  }
  if (product.onSale) {
    return `⚡${product.name} - ${product.originalPrice}원 → ${product.price}원 (20% SALE!)`;
  }
  if (product.suggestSale) {
    return `💝${product.name} - ${product.originalPrice}원 → ${product.price}원 (5% 추천할인!)`;
  }
  return `${product.name} - ${product.price}원`;
};
```
이런식으로 string을 바로 return 시키게 구성하고 나서
```js
const getProductDisplayText = (product: Product): string => {
  const text = getPriceText();
  return product.quantity === 0 ? `${text} (품절)` : text;
};
```
이렇게 const 로 하나의 텍스트로 바로 배출할수 있게끔 가독성 확장성을 높힐수 있을 것 같아요

unseoJang commented on src/advanced/components/Manual.tsx:
이부분 컴포넌트 조금만 더 상세하게 나눌수 잇을 것같아요!
응집도 부분이 걱정된다면 이번 챕터에서 다 배울수 잇을 것같으니 넘 걱정 마시고
혹시나 도움이 되거나 참고하실 부분이 있으면 이번 과제 현지수님이 참 잘해주셔서
현지수님 과제를 참고해도 좋을 것 같아요!

unseoJang commented on src/advanced/components/Toast.tsx:
```diff
@@ -0,0 +1,71 @@
+    return () => clearTimeout(timer);
+  }, [message.id, message.duration, onRemove]);
+
+  const getToastStyles = () => {
```
스타일 class 분기처리 해주신 부분 좋네요

ckdwns9121 committed:
fix: 불필요한 query 로직 삭제

unseoJang commented on src/advanced/context/CartContext.tsx:
```diff
@@ -0,0 +1,212 @@
+  selectedProductId: null,
+};
+
+const cartReducer = (state: AppState, action: CartAction): AppState => {
```
생각보다 해당 함수가 많이 무거운 듯한 느낌?
가독성이 많이 떨어진 다는 느낌을 받앗어요
세부 함수들 헬퍼를 이용해 분기점 나눠주면 확장성 가독성이 더 올라갈것 같아요
```js
function updateProductStock(products, productId, quantityDelta) {
  return products.map(p =>
    p.id === productId ? { ...p, quantity: p.quantity + quantityDelta } : p
  );
}

function updateCartItemQuantity(items, productId, change) {
  return items.map(item =>
    item.id === productId ? { ...item, quantity: item.quantity + change } : item
  );
}
```
이런식으로요

ckdwns9121 replied:
오 되게 좋은것 같습니다. 안그래도 코드가 길어지고 가독성이 떨어지는 함수라고 생각하고 있었긴 한데 좋은 리뷰 감사드려요!
바로 고쳐보겠습니다 ㅎㅎ

ckdwns9121 replied:
```tsx

const handleRemoveFromCart = (state: AppState, productId: string): AppState => {
  const item = state.cartItems.find(item => item.id === productId);
  if (!item) return state;

  const updatedProducts = state.products.map(p => (p.id === productId ? { ...p, quantity: p.quantity + item.quantity } : p));

  return {
    ...state,
    cartItems: state.cartItems.filter(item => item.id !== productId),
    products: updatedProducts,
  };
};

const handleClearCart = (state: AppState): AppState => {
  // 모든 아이템을 재고로 반환
  const updatedProducts = state.products.map(product => {
    const cartItem = state.cartItems.find(item => item.id === product.id);
    return cartItem ? { ...product, quantity: product.quantity + cartItem.quantity } : product;
  });

  return {
    ...state,
    cartItems: [],
    products: updatedProducts,
  };
};
```

이런식으로 헬퍼함수를 분리하고

```tsx
const cartReducer = (state: AppState, action: CartAction): AppState => {
  switch (action.type) {
    case "ADD_TO_CART":
      return handleAddToCart(state, action.productId);

    case "UPDATE_QUANTITY":
      return handleUpdateQuantity(state, action.productId, action.change);

    case "REMOVE_FROM_CART":
      return handleRemoveFromCart(state, action.productId);

    case "CLEAR_CART":
      return handleClearCart(state);

    case "UPDATE_PRODUCTS":
      return handleUpdateProducts(state, action.products);

    case "SET_SELECTED_PRODUCT":
      return handleSetSelectedProduct(state, action.productId);

    default:
      return state;
  }
};
```

최종적으로는 reducer가 내부 로직은 모르게끔 추상화해서 이런식으로 구현해봤는데 괜찮은거같네요 !!

unseoJang commented on src/basic/services/timerService.js:
```diff
@@ -0,0 +1,122 @@
+  // 추천세일 타이머 시작
+  startSuggestSaleTimer() {
+    const suggestDelay = Math.random() * 20000;
+    setTimeout(() => {
```
```js
export const startRandomlyInMs = (ms: number) => (callback: () => void) => {
  const startsAt = Math.random() * ms;
  return setTimeout(callback, startsAt);
};
```
저번기수 도형님이라는 분의 코드인데요 이렇게 까지 극단적으로 구조화 하시는 분도 계셧어요

ckdwns9121 replied:
캬 타이머 생성하는것도 유틸함수로 분리하신분이 있군요.. 좋은거같아요

unseoJang commented on src/basic/services/productService.js:
```diff
@@ -0,0 +1,191 @@
+/* eslint-disable indent */
```
??eslint 왜 껏는지 알수 있을까요?

ckdwns9121 committed:
refactor: 서비스 초기화 로직 분리

ckdwns9121 committed:
fix: 들여쓰기 eslint 수정

unseoJang commented on src/advanced/context/CartContext.tsx:
```diff
@@ -0,0 +1,212 @@
+      };
+    }
+
+    case "SET_SELECTED_PRODUCT": {
```
case에 들어가잇는 string 값들 변하지 않는 값이니 상수화해도 좋을 것 같아요

unseoJang commented on src/advanced/context/CartContext.tsx:
cartReducer의 코드 가 개인적으로 많이 무겁다고 느껴지네요
안에 내용들 좀더 구조 나눠줄수 있을 것 같아요

ckdwns9121 committed:
refactor: 서비스 이름을 동적으로 return

unseoJang commented on src/basic/services/productService.js:
현재 class 마다 너무 많은 역할이 부여되어있는 것이 느껴져요
제가 class보단 function을 더 선호하긴하지만
```js
ProductService = {
  getQueryHelper() => ProductQuery
  getStockHelper() => StockService
  getDiscountHelper() => SaleService
}
```
이렇게 관심사 별로 로직을 분리하면 디버깅 하기도 쉬워지고 한 class가 많은 책임을 담당하는 것을 막아줄수 있다고 생각이드네요

두번째로 여기도 비슷한 Object를 return 시켜주는 모습이 보이네요

```js
return {
            ...product,
            price: Math.round(product.originalPrice * DISCOUNT_RATES.LIGHTNING_SALE),
            onSale: true,
          };
```
```js
    return {
            ...product,
            price: Math.round(product.price * DISCOUNT_RATES.SUGGEST_SALE),
            suggestSale: true,
          };
        }
  ```

분명 하나로 묶어서 관리해줄수 있다고 생각이 드는 부분이기도 하고 안에 들어가 있는 계산식도 메서드 하나로 묶어서 정리해주면 더 깔끔한 코드가 될것 같습니다.

ckdwns9121 replied:
좋은거같아요. 해당 리팩토링은 Command-Query Separation 원칙과 관련이 있을거 같은데 적절히 Command 함수랑 Query함수를 분리해 헬퍼로 만들어서 책임을 관리해도 좋을거같다는 생각이드네요..!

unseoJang commented on src/basic/services/timerService.js:
```diff
@@ -0,0 +1,122 @@
+    setTimeout(() => {
+      this.lightningSaleTimer = setInterval(() => {
+        this.applyLightningSale();
+      }, 30000);
```
이 30000이라는 초 단위도 상수화 시켜도 좋을 것 같아요

unseoJang commented on src/basic/services/timerService.js:
```diff
@@ -0,0 +1,122 @@
+
+  // 번개세일 타이머 시작
+  startLightningSaleTimer() {
+    const lightningDelay = Math.random() * 10000;
```
과제를 극단적으로 경험해보려면 이런 식도 아예 함수로 빼보면 연습에 도움이 될것같아요

unseoJang commented on src/basic/services/timerService.js:
```diff
@@ -0,0 +1,122 @@
+    if (result.success) {
+      alert("💝 " + result.product.name + "은(는) 어떠세요? 지금 구매하시면 5% 추가 할인!");
+      this.notifyUIUpdate(); // ✅ UI 업데이트를 이벤트로 분리
+      console.log(result.message);
```
클린코드 주차기 때문에 console.log 는 깔끔히 삭제~

unseoJang commented on src/basic/services/timerService.js:
```diff
@@ -0,0 +1,122 @@
+    const result = this.productService.applySuggestSale(lastSelectedProduct?.id);
+
+    if (result.success) {
+      alert("💝 " + result.product.name + "은(는) 어떠세요? 지금 구매하시면 5% 추가 할인!");
```
해당  string 도 상수화 시켜서 딸 관리 해주면 좋을 것 같아요

unseoJang commented on src/basic/services/timerService.js:
중간 중간 비슷한 object를 리턴 시켜주는 모습이 보여요
하나로 합쳐서 상단에서 관리해주고 return 시켜주면 훨씬 깔끔한 코드가 될것 같아요

unseoJang commented on src/basic/utils/discountUtils.js:
```diff
@@ -0,0 +1,43 @@
+export function calcTotalDiscount(individualDiscounts, bulkDiscountRate, bulkDiscountApplied, tuesdayDiscountRate, tuesdayDiscountApplied) {
```
이 함수에서 let을 사용하고 있는 이유는 totalRate와 discountTypes를 조건에 따라 누적/수정하기 위함인것 같은데, 조건 분기를 map → filter → reduce 등으로 함수형 처리하면 let 없이도 훨씬 클린하게 표현할 수 있습니다.

```js
export function calcTotalDiscount(
  individualDiscounts,
  bulkDiscountRate,
  bulkDiscountApplied,
  tuesdayDiscountRate,
  tuesdayDiscountApplied
) {
  const baseDiscounts = individualDiscounts.map(discount => ({
    rate: discount.rate,
    type: "individual",
  }));

  const optionalDiscounts = [
    bulkDiscountApplied && { rate: bulkDiscountRate, type: "bulk" },
    tuesdayDiscountApplied && { rate: tuesdayDiscountRate, type: "tuesday" },
  ].filter(Boolean); // false/null 제거

  const allDiscounts = [...baseDiscounts, ...optionalDiscounts];

  const totalDiscountRate = allDiscounts.reduce((sum, d) => sum + d.rate, 0);
  const discountTypes = allDiscounts.map(d => d.type);

  return {
    totalDiscountRate,
    discountTypes,
    hasAnyDiscount: totalDiscountRate > 0,
  };
}
```

ckdwns9121 replied:
오 함수형 패러다임을 적용해서 pipe 식으로 한번 만들어볼께요

unseoJang commented on src/advanced/components/Layout.tsx:
이 부분 컴포넌트로 좀 더 묶어줄수 있는 부분이 있을 것같아요
그리고 과제다 보니 해당 과제를 극단적으로 해보려면 className 에 들어간 css styling name 도 상수화 시켜보면 연습에 도움이 될것같아요

unseoJang commented on src/advanced/constants/index.ts:
상수화 잘 해주셧네요 ts의 enum값도 사용해보면 좋을 것 같아요

unseoJang commented on src/advanced/context/ToastContext.tsx:
중간중간 주석이 좀 필요해보여요

unseoJang commented on src/advanced/hooks/useSpecialEvents.ts:
```diff
@@ -0,0 +1,85 @@
+    const randomIndex = Math.floor(Math.random() * availableProducts.length);
+    const luckyProduct = availableProducts[randomIndex];
+
+    const newPrice = Math.round((luckyProduct.originalPrice * (100 - DISCOUNT_RATES.LIGHTNING_DISCOUNT)) / 100);
```
이런 식 자체도 함수로 분리해주면 좋고 안에 들어가는 숫자들도 상수로 따로 관리해주면 가독성 올라가고 다른사람들과 협업할떄 관리하기 편할것같아요

unseoJang commented on src/advanced/hooks/useSpecialEvents.ts:
```diff
@@ -0,0 +1,85 @@
+      onSale: true,
+    });
+
+    addToast({
```
동일한 객체는 상단에서 같이 관리해주면 좋을 것 같아요!

unseoJang reviewed (commented):
안녕하세요 창준님~
2-1과제 진행하시느라 고생이 많으셨습니다.
이번주차 과제에서 아마 느낀게 많으실것같아요 ㅎㅎ
각 파일에서 함수들이 맡고있는 역할들이 진짜 큰것같아요
이번 주차 과제를 극단적으로 조그마한 단위별로 구조분리 연습을 해본다면 창준님에게 큰 도움이 될것같아요.
이번주차 과제 지수님이 구조분리 네이밍 상수화 도메인별 분리 등등 이미 2챕터 과제 전반적으로 진행을 많이 해주셧어요 한번 참고하면 창준님께 도움이 될것 같아요
이번주 과제는 극단적으로 해야지 담번에 실제 실무로 가면 도움이 되는 그런 과제라고 생각이들어요
이번주 과제 정말 수고많으셧고 2-2과제도 화이팅입니다!👍

ckdwns9121 committed:
refactor: 불필요한 let 제거

ckdwns9121 committed:
refactor: getPriceText 함수를 분리

ckdwns9121 committed:
refactor: Manual 컴포넌트를 조금더 세분화

ckdwns9121 committed:
refactor: ProductService 관심사 분리 및 중복 코드 제거

- ProductQueryHelper, StockHelper, SaleHelper로 관심사 분리
- ProductUpdateHelper로 상품 업데이트 패턴 통합
- 할인 계산 로직을 calculateDiscountedPrice로 통합
- 각 헬퍼에 대한 접근자 메서드 추가 (getQueryHelper, getStockHelper, getSaleHelper)
- 단일 책임 원칙 적용으로 유지보수성 및 테스트 용이성 향상

ckdwns9121 committed:
refactor: calcTotalDiscount를 함수형 프로그래밍으로 개선

- let 제거하고 filter/reduce/map 사용
- 명령형 → 함수형 패턴으로 변경

ckdwns9121 committed:
refactor: ToastContainer 및 CartContext의 코드 정리 및 주석 추가

- ToastContainer 컴포넌트에 Props 인터페이스 및 설명 추가
- CartContext의 장바구니 관련 로직을 헬퍼 함수로 분리하여 가독성 향상
- 각 헬퍼 함수에 대한 설명 추가로 코드 이해도 개선

ckdwns9121 committed:
refactor: useSpecialEvents 코드 개선 및 중복 제거

- 공통 토스트 설정 및 상품 업데이트 헬퍼 함수 추가로 코드 가독성 향상
- 중복된 상품 업데이트 및 토스트 생성 로직을 헬퍼 함수로 통합하여 유지보수성 개선

ckdwns9121 committed:
refactor: useSpecialEvents에서 토스트 생성 로직 개선

- createToast 함수를 객체 구조 분해 할당을 사용하여 가독성 향상
- addToast 호출 시 createToast에 객체를 전달하여 코드 일관성 및 유지보수성 개선

ckdwns9121 committed:
refactor: DiscountService 코드 구조 개선 및 헬퍼 함수 추가

- 할인 계산 로직을 헬퍼 객체로 분리하여 가독성 향상
- 개별 상품 할인 계산, 총계 계산, 대량구매 및 화요일 할인 적용 로직을 헬퍼 함수로 통합
- DiscountService의 메서드에서 헬퍼 객체를 통해 할인 계산 로직을 호출하도록 변경하여 유지보수성 개선

ckdwns9121 committed:
refactor: OrderService에 주문 계산 헬퍼 추가

- 주문 계산 관련 로직을 OrderCalculationHelper 객체로 분리하여 가독성 향상
- 할인율, 아이템 수량, 화요일 여부 확인 및 할인 정보 포맷팅 기능 추가
- OrderService에서 헬퍼 객체를 통해 계산 로직을 호출하도록 변경하여 유지보수성 개선

ckdwns9121 committed:
refactor: 장바구니 서비스에 헬퍼 함수 추가 및 코드 구조 개선

- 장바구니 관련 로직을 여러 헬퍼 객체로 분리하여 가독성 향상
- 상품 유효성 검증, 재고 검증, 장바구니 계산 및 상태 관리 기능을 헬퍼로 통합
- CartService에서 헬퍼 객체를 통해 로직을 호출하도록 변경하여 유지보수성 개선
